ts/sagas/identification.ts

```diff
@@ -0,0 +1,131 @@
+import { Effect } from "redux-saga";
+import { call, put, select, take, takeLatest } from "redux-saga/effects";
+import { ActionType, getType } from "typesafe-actions";
+
+import { startApplicationInitialization } from "../store/actions/application";
+import { sessionInvalid } from "../store/actions/authentication";
+import {
+  identificationCancel,
+  identificationPinReset,
+  identificationRequest,
+  identificationReset,
+  identificationStart,
+  identificationSuccess
+} from "../store/actions/identification";
+import { navigateToMessageDetailScreenAction } from "../store/actions/navigation";
+import { clearNotificationPendingMessage } from "../store/actions/notifications";
+import {
+  IdentificationCancelData,
+  IdentificationResult,
+  IdentificationSuccessData
+} from "../store/reducers/identification";
+import {
+  PendingMessageState,
+  pendingMessageStateSelector
+} from "../store/reducers/notifications/pendingMessage";
+import { GlobalState } from "../store/reducers/types";
+import { PinString } from "../types/PinString";
+import { SagaCallReturnType } from "../types/utils";
+import { deletePin } from "../utils/keychain";
+
+// Wait the identification and return the result
+export function* waitIdentificationResult(): Iterator<
+  Effect | IdentificationResult
+> {
+  const resultAction:
+    | ActionType<typeof identificationCancel>
+    | ActionType<typeof identificationPinReset>
+    | ActionType<typeof identificationSuccess> = yield take([
+    getType(identificationCancel),
+    getType(identificationPinReset),
+    getType(identificationSuccess)
+  ]);
+
+  // If the identification was cancelled just return false
+  if (resultAction.type === getType(identificationCancel)) {
+    return IdentificationResult.cancel;
+  }
+
+  // If the user decided to reset the pin perform needed actions than return false
+  if (resultAction.type === getType(identificationPinReset)) {
+    // Delete the PIN
+    // tslint:disable-next-line:saga-yield-return-type
+    yield call(deletePin);
```

Comment:
I would delete the pin after invalidating the session: say there is a bug that causes the app to crash just after deleting the pin, on the next run, the session is still valid but there's no pin, the user gets asked to set a new pin...